// unplugin-lezer/dist/chunk-HLXNQ6LD.js
// src/index.ts
import { readFile } from "fs/promises";
import { dirname, resolve } from "path";
import process from "process";
import { createUnplugin } from "unplugin";
import { buildParserFile } from "@lezer/generator";
var unpluginFactory = () => ({
  name: "unplugin-lezer",
  resolveId(source, importer) {
    const m = /^(.*\.grammar)(\.terms)?$/.exec(source);
    if (!m) return null;
    const id = resolve(importer ? dirname(importer) : process.cwd(), m[1]);
    return m[2] ? `\0${id}.terms` : id;
  },
  load(id) {
    const m = /^\0?(.*\.grammar)(\.terms)?$/.exec(id);
    if (!m) return null;
    if (!m[2]) this.addWatchFile(id);
    const base = m[1];
    const build = readFile(base, "utf8").then(
      (code) => buildParserFile(code, {
        fileName: base,
        moduleStyle: "es",
        warn: (message) => this.warn(message)
      })
    );
    return build.then(
      (result) => (m == null ? void 0 : m[2]) ? result.terms : result.parser
    );
  }
});
var unplugin = /* @__PURE__ */ createUnplugin(unpluginFactory);
var src_default = unplugin;

export {
  unpluginFactory,
  unplugin,
  src_default
};
