// unplugin-lezer/dist/esbuild.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/esbuild.ts
var esbuild_exports = {};
__export(esbuild_exports, {
  default: () => esbuild_default
});
module.exports = __toCommonJS(esbuild_exports);
var import_unplugin2 = require("unplugin");

// src/index.ts
var import_promises = require("fs/promises");
var import_path = require("path");
var import_process = __toESM(require("process"), 1);
var import_unplugin = require("unplugin");
var import_generator = require("@lezer/generator");
var unpluginFactory = () => ({
  name: "unplugin-lezer",
  resolveId(source, importer) {
    const m = /^(.*\.grammar)(\.terms)?$/.exec(source);
    if (!m) return null;
    const id = (0, import_path.resolve)(importer ? (0, import_path.dirname)(importer) : import_process.default.cwd(), m[1]);
    return m[2] ? `\0${id}.terms` : id;
  },
  load(id) {
    const m = /^\0?(.*\.grammar)(\.terms)?$/.exec(id);
    if (!m) return null;
    if (!m[2]) this.addWatchFile(id);
    const base = m[1];
    const build = (0, import_promises.readFile)(base, "utf8").then(
      (code) => (0, import_generator.buildParserFile)(code, {
        fileName: base,
        moduleStyle: "es",
        warn: (message) => this.warn(message)
      })
    );
    return build.then(
      (result) => (m == null ? void 0 : m[2]) ? result.terms : result.parser
    );
  }
});

// src/esbuild.ts
var esbuild_default = (0, import_unplugin2.createEsbuildPlugin)(unpluginFactory);
exports.default = module.exports;
